### ISSUE TITLE
Demographics and Units bug

### ISSUE THREAD

**Issue by daniel-klein** [starsimhub/starsim#732]
Currently, if a user specifies units other than `year`, the demographic modules break.

```python
Exception has occurred: ValueError
could not broadcast input array from shape (209,) into shape (1,)
  File "/Users/dklein/GIT/starsim/starsim/arrays.py", line 83, in __setitem__
    self.values[index] = value
    ~~~~~~~~~~~^^^^^^^
  File "/Users/dklein/GIT/starsim/starsim/demographics.py", line 252, in finalize
    self.results.cmr[:] = 1/self.pars.rate_units * np.divide(self.results.new / self.sim.dt_year, n_alive, where=n_alive>0)
    ~~~~~~~~~~~~~~~~^^^
  File "/Users/dklein/GIT/starsim/starsim/sim.py", line 360, in finalize
    module.finalize()
  File "/Users/dklein/GIT/starsim/starsim/sim.py", line 341, in run
    self.finalize()
  File "/Users/dklein/GIT/tbsim/scripts/general/run_tb.py", line 42, in <module>
    sim_tb.run()
ValueError: could not broadcast input array from shape (209,) into shape (1,)
```

Because:
```python
n_alive
Result(sim.n_alive):
array([999. 990. 988. 980. 975. 966. 960. 953. 947. 943. 938. 930. 924. 917.
 911. 908. 899. 890. 880. 871. 866. 857. 849. 835. 822. 813. 804. 792.
 783. 771. 763. 755. 747. 738. 733. 719. 712. 702. 692. 684. 670. 663.
 654. 644. 638. 632. 628. 625. 618. 611. 606. 603. 599. 595. 588. 581.
 580. 574. 571. 569. 566. 563. 559. 557. 555. 552. 547. 541. 538. 537.
 536. 536. 536. 532. 529. 528. 526. 523. 523. 522. 521. 521. 518. 517.
 516. 513. 513. 512. 510. 509. 504. 503. 502. 502. 501. 500. 500. 500.
 500. 497. 497. 496. 496. 495. 494. 494. 492. 491. 491. 491. 491. 491.
 490. 490. 489. 489. 489. 489. 489. 487. 487. 487. 487. 486. 486. 485.
 484. 484. 484. 484. 484. 484. 484. 483. 483. 483. 483. 483. 483. 483.
 482. 482. 481. 481. 481. 481. 481. 481. 481. 481. 481. 481. 481. 481.
 481. 481. 481. 481. 481. 481. 480. 480. 479. 479. 479. 479. 479. 479.
 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479.
 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479. 479.
 479. 479. 479. 479. 479. 479. 479. 479. 479. 478. 478. 478. 478.])
self.results.new
Result(deaths.new):
array([92.])
```

See also #709

**Comment by daniel-klein** [comment]
And worse, if demographics modules are given a unit other than the default `year`, I believe the calculations are wrong. Again see #709.

**Comment by daniel-klein** [comment]
Related: add more tests involving units.

**Comment by cliffckerr** [comment]
This has ended up being pretty gnarly -- the title of this blog says it all :D
https://ondata.blog/articles/nine-circles-of-hell-time-in-python/

Woring on it in #733.

**Comment by cliffckerr** [comment]
Fixed finally, but slight sim slowdown (10%) -- see #744. Closed by #724
